AmbiguousResolutionException when same class is present twice and certain features are used

**Describe the bug**  
When a class is present twice in an application and is used as a CDI managed bean and certain features are used, CDI may sometimes report an `AmbiguousResolutionException`.

```
Caused by: org.jboss.weld.exceptions.AmbiguousResolutionException: WELD-001318: Cannot resolve an ambiguous dependency between: 
  - Managed Bean [class com.ibm.wssvt.acme.annuity.common.business.jaxrs.MyJAXBResolver] with qualifiers [@Any @Default],
  - Managed Bean [class com.ibm.wssvt.acme.annuity.common.business.jaxrs.MyJAXBResolver] with qualifiers [@Any @Default]
```

**Steps to Reproduce**
- Enable the mpFaultTolerance feature (any version)
- Create an EAR containing a .jar in the lib directory and a .war.
- Place the same CDI managed bean class inside both the .jar and the .war
- Note that the class can be loaded fine (you will always load it from the .jar)
- Attempt to inject and call the CDI managed bean
- An exception is thrown, reporting that the bean cannot be resolved because it is ambiguous

Note: the exception depends on the exact order that CDI scans the .jar and the .war, so the error does not happen every time.

**Expected behavior**  
- Assuming normal parent first classloading, all parts of the application will always see the version of the class from the .jar
- Therefore CDI should ignore the class in the .war
- There should not be a bean resolution exception

**Diagnostic information:**  
 - OpenLiberty Version: all
 - Affected feature(s): `cdi-1.2` or later  along with one or more of `mpFaultTolerance` (all versions), `mpGraphQL` (allVersions), `mpReactiveMessaging-1.0`, `mpRestClient-3.0` or `appSecurity-5.0`
 - Java Version: n/a
 - server.xml configuration: n/a
 - If it would be useful, upload the messages.log file found in `$WLP_OUTPUT_DIR/messages.log`

**Additional context**  
Add any other context about the problem here.

There are assumptions in the CDI code that more widely visible parts of the application (e.g. ear libs) are scanned before less widely visible parts (e.g. war classes). When a feature which includes a CDI extension which has visibility of all application BDAs is included, these assumptions are violated and the same bean can be incorrectly included twice.

--- a/dev/com.ibm.ws.cdi.visibility_fat/bnd.bnd
+++ b/dev/com.ibm.ws.cdi.visibility_fat/bnd.bnd
@@ -37,7 +37,9 @@ tested.features:\
 	cdi-4.0,\
 	connectors-2.1,\
 	jakartaee-10.0,\
-	servlet-6.0
+	servlet-6.0,\
+	mpConfig-3.0,\
+	mpFaultTolerance-4.0
 
 -buildpath: \
 	com.ibm.websphere.javaee.ejb.3.2;version=latest,\

--- a/dev/com.ibm.ws.cdi.visibility_fat/fat/src/com/ibm/ws/cdi/visibility/tests/ejb/EJBVisibilityTests.java
+++ b/dev/com.ibm.ws.cdi.visibility_fat/fat/src/com/ibm/ws/cdi/visibility/tests/ejb/EJBVisibilityTests.java
@@ -37,7 +37,7 @@ import componenttest.annotation.TestServlets;
 import componenttest.custom.junit.runner.FATRunner;
 import componenttest.custom.junit.runner.Mode;
 import componenttest.custom.junit.runner.Mode.TestMode;
-import componenttest.rules.repeater.EERepeatActions;
+import componenttest.rules.repeater.MicroProfileActions;
 import componenttest.rules.repeater.RepeatTests;
 import componenttest.topology.impl.LibertyServer;
 import componenttest.topology.utils.FATServletClient;
@@ -49,7 +49,7 @@ public class EJBVisibilityTests extends FATServletClient {
     public static final String SERVER_NAME = "cdi12EJBServer";
 
     @ClassRule
-    public static RepeatTests r = EERepeatActions.repeat(SERVER_NAME, EERepeatActions.EE10, EERepeatActions.EE9, EERepeatActions.EE7);
+    public static RepeatTests r = MicroProfileActions.repeat(SERVER_NAME, MicroProfileActions.MP60, MicroProfileActions.MP14, MicroProfileActions.MP50);
 
     public static final String MASKED_CLASS_APP_NAME = "maskedClassWeb";
 

--- a/dev/com.ibm.ws.cdi.visibility_fat/publish/servers/cdi12EJBServer/server.xml
+++ b/dev/com.ibm.ws.cdi.visibility_fat/publish/servers/cdi12EJBServer/server.xml
@@ -10,6 +10,9 @@
         <feature>localConnector-1.0</feature>
         <feature>appClientSupport-1.0</feature>
         <feature>componentTest-1.0</feature>
+        <!-- Fault Tolerance included as a feature with CDI extension which can see application BDAs -->
+        <!-- which caused issues with masked classes -->
+        <feature>mpFaultTolerance-1.1</feature>
     </featureManager>
     
     <orb id="defaultOrb" orbSSLInitTimeout="60"/>

--- a/dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/weld/BeanDeploymentArchiveImpl.java
+++ b/dev/com.ibm.ws.cdi.weld/src/com/ibm/ws/cdi/impl/weld/BeanDeploymentArchiveImpl.java
@@ -255,10 +255,14 @@ public class BeanDeploymentArchiveImpl implements WebSphereBeanDeploymentArchive
             //mark as scanned up front to prevent loops
             this.scanned = true;
 
-            //scan the children
-            for (WebSphereBeanDeploymentArchive child : accessibleBDAs) {
-                if (!child.hasBeenScanned()) {
-                    child.scan();
+            // Scan any accessible BDAs first to ensure we scan more visible things (shared libs, ear libs) before less visible things (war classes, war libs)
+            // This helps to make sure that the later call to isAccessibleBean works
+            // Don't scan accessible BDAs of runtime extensions because they sit outside the hierarchy and some of them need to see everything
+            if (getType() != ArchiveType.RUNTIME_EXTENSION) {
+                for (WebSphereBeanDeploymentArchive child : accessibleBDAs) {
+                    if (!child.hasBeenScanned()) {
+                        child.scan();
+                    }
                 }
             }
 

